tools/server/server.cpp

```diff
@@ -4332,9 +4332,69 @@ int main(int argc, char ** argv) {
         }
     };
 
-    const auto handle_completions = [&handle_completions_impl](const httplib::Request & req, httplib::Response & res) {
+    const auto handle_completions = [&ctx_server, &handle_completions_impl](const httplib::Request & req, httplib::Response & res) {
         json data = json::parse(req.body);
-        std::vector<raw_buffer> files; // dummy
+        json medias = json_value(data, "medias", json::array());
+        auto & opt = ctx_server.oai_parser_opt;
+        std::vector<raw_buffer> files;
+
+        if (medias.is_array()) {
+            for (auto & m : medias) {
+                std::string type = json_value(m, "type", std::string());
+                std::string data = json_value(m, "data", std::string());
+                if (type.empty() || data.empty()) {
+                    continue;
+                }
+                if (type == "image_url" || type == "image" || type == "img") {
+                    if (!opt.allow_image) {
+                        throw std::runtime_error("image input is not supported - hint: if this is unexpected, you may need to provide the mmproj");
+                    }
+                    if (string_starts_with(data, "http")) {
+                        // download remote image
+                        common_remote_params params;
+                        params.headers.push_back("User-Agent: llama.cpp/" + build_info);
+                        params.max_size = 1024 * 1024 * 10; // 10MB
+                        params.timeout  = 10; // seconds
+                        SRV_INF("downloading image from '%s'\n", data.c_str());
+                        auto res = common_remote_get_content(data, params);
```

Comment:
Instead of duplicating this whole code block, extract it to a general function and reuse it in `/chat/completion` and `/completion`. [DRY code principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)